li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaInstrumentedProducerImpl.java

```diff
@@ -460,36 +459,39 @@ private boolean proceedClosing() {
     if (isClosed()) {
       return false;
     }
-    synchronized (closeLock) {
-      if (isClosed()) {
-        return false; //someone beat us to it
-      }
-      int holds = delegateLock.getReadHoldCount(); //this is for our thread
-      ReentrantReadWriteLock.ReadLock readLock = delegateLock.readLock();
-      ReentrantReadWriteLock.WriteLock writeLock = delegateLock.writeLock();
-      if (holds > 0) { //do we own a read lock ?
-        for (int i = 0; i < holds; i++) {
-          readLock.unlock();
-        }
-        //at this point we no longer hold a read lock, but any number of other
-        //readers/writers may slip past us
+
+    // release read lock held by current thread if any
+    // BMM has use cases of closing the producer in send callback if send fails, we need to ensure
```

Comment:
nit - not specific to bmm. closing from a callback is sadly a kafka feature

Reply:
Is there a way to disable that? Of course I am not suggesting we do this in this PR.